File: exercise/php/day13/src/SantaMarket/ReceiptItem.php
<?php

namespace SantaMarket;

class ReceiptItem
{
    private Product $product;
    private float $price;
    private float $totalPrice;
    private float $quantity;

    public function __construct(Product $product, float $quantity, float $price, float $totalPrice)
    {
        $this->product = $product;
        $this->quantity = $quantity;
        $this->price = $price;
        $this->totalPrice = $totalPrice;
    }

    public function equals(ReceiptItem $other, float $tolerance = 0.001): bool
    {
        return $this->product->equals($other->getProduct()) &&
            $this->isCloseEnough($this->price, $other->getPrice(), $tolerance) &&
            $this->isCloseEnough($this->totalPrice, $other->getTotalPrice(), $tolerance) &&
            $this->isCloseEnough($this->quantity, $other->getQuantity(), $tolerance);
    }

    public function getProduct(): Product
    {
        return $this->product;
    }

    private function isCloseEnough(float $a, float $b, float $tolerance): bool
    {
        return abs($a - $b) <= $tolerance;
    }

    public function getPrice(): float
    {
        return $this->price;
    }

    public function getTotalPrice(): float
    {
        return $this->totalPrice;
    }

    public function getQuantity(): float
    {
        return $this->quantity;
    }
}

File: solution/day13/src/main/java/santamarket/model/ReceiptItem.java
package santamarket.model;

import com.google.common.math.DoubleMath;
import java.util.Objects;

public class ReceiptItem {
    private final Product product;
    private final double price;
    private final double totalPrice;
    private final double quantity;

    ReceiptItem(Product p, double quantity, double price, double totalPrice) {
        this.product = p;
        this.quantity = quantity;
        this.price = price;
        this.totalPrice = totalPrice;
    }

    public double getPrice() {
        return price;
    }

    public Product getProduct() {
        return product;
    }

    public double getQuantity() {
        return quantity;
    }

    public double getTotalPrice() {
        return totalPrice;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ReceiptItem)) return false;
        ReceiptItem that = (ReceiptItem) o;
        double tolerance = 0.001;
        return DoubleMath.fuzzyCompare(that.price, price, tolerance) == 0 &&
                DoubleMath.fuzzyCompare(that.totalPrice, totalPrice, tolerance) == 0 &&
                DoubleMath.fuzzyCompare(that.quantity, quantity, tolerance) == 0 &&
                Objects.equals(product, that.product);
    }

    @Override
    public int hashCode() {
        return Objects.hash(product, price, totalPrice, quantity);
    }

    @Override
    public String toString() {
        return "ReceiptItem{" +
                "product=" + product +
                ", price=" + price +
                ", totalPrice=" + totalPrice +
                ", quantity=" + quantity +
                '}';
    }
}
